### ISSUE TITLE
Scope errors with "websocket: close sent" and web UI not refreshed

### ISSUE THREAD

**Issue by alban** [weaveworks/scope#1066]
I created a Kubernetes cluster based on `cluster.yaml` from https://coreos.com/kubernetes/docs/latest/kubernetes-on-aws.html#kube-aws-quickstart:

Then, I install Scope on Kubernetes with yaml files inspired by [weavescope](https://github.com/TheNewNormal/kube-charts/tree/master/weavescope/manifests)
- [scope-app-rc.yaml](https://gist.github.com/alban/4755128c677bd1f42f75)
- [scope-app-svc.yaml](https://gist.github.com/alban/995babe4c17bdf63ea40)
- [scope-probe-ds.yaml](https://gist.github.com/alban/f2d402c2a288571db640)

I am also installing some pods based on [guestbook-all-in-one.yaml](https://github.com/kubernetes/kubernetes/blob/master/examples/guestbook/all-in-one/guestbook-all-in-one.yaml).

I am using `ssh -L 4042:10.3.0.131:4040` and `http://localhost:4042/` to connect to scope app service from my laptop. I can see my containers in the web UI but after a couple of clicks to explore, the web UI is not refreshed: the "load in progress" animated icon is spinning. And I can see the following logs in Kubernetes:

```
$ kubectl --kubeconfig=clusters/demo-kubecon-1/kubeconfig logs weave-scope-app-rdiez
<app> INFO: 2016/02/29 17:11:49.412306 app starting, version 3e611f8, ID 7ad8e3ccecaa8e9e
<app> INFO: 2016/02/29 17:11:49.454402 listening on :4040
<app> ERRO: 2016/02/29 17:11:49.454679 Error updating weaveDNS, backing off 20s: Error running weave expose: exit status 1
<app> ERRO: 2016/02/29 17:12:09.494532 Error updating weaveDNS, backing off 40s: Error running weave expose: exit status 1
<app> ERRO: 2016/02/29 17:12:49.544389 Error updating weaveDNS, backing off 1m0s: Error running weave expose: exit status 1
<app> ERRO: 2016/02/29 17:15:23.538895 cannot serialize topology diff: websocket: close sent
```

I also check the logs on probes:

```
$ kubectl --kubeconfig=clusters/demo-kubecon-1/kubeconfig logs weave-scope-probe-ape8m
<probe> INFO: 2016/02/29 17:12:51.231727 probe starting, version 3e611f8, ID 9bb50c53ed8315a
<probe> INFO: 2016/02/29 17:12:51.231812 publishing to: 10.3.0.131:4040
<probe> INFO: 2016/02/29 17:12:51.336866 docker container: collecting stats for df128ffac627572e57417f365514176ff26a9eed9abb7d648bf458928a848e79
<probe> INFO: 2016/02/29 17:12:51.343960 Control connection to 10.3.0.131:4040 starting
<probe> INFO: 2016/02/29 17:12:51.491248 Publish loop for 10.3.0.131:4040 starting
<probe> INFO: 2016/02/29 17:12:51.491859 docker container: collecting stats for 0c7237621cc936084e5c5d438ebe063d21556e138280821833054d4f1111a948
<probe> INFO: 2016/02/29 17:12:51.548904 docker container: collecting stats for b0037300c2a78a483a85ade35a24743147cc62eece90427b7e0cca7725ed545a
<probe> INFO: 2016/02/29 17:12:51.584889 docker container: collecting stats for 7c5ee4e080ff6f0512d59eea5f49e37f048598d40a637077d0c78e9fcf39a32e
<probe> INFO: 2016/02/29 17:12:51.633982 docker container: collecting stats for b8c24f7492f56f6721e2d455bf440537d4b544aba5a4d028b681e6cdd3cb7606
<probe> INFO: 2016/02/29 17:12:51.689417 docker container: collecting stats for 5263b865ab4a0212f62c0c4e1579cc10032fdc6af456654374f7d41ad1e44b01
<probe> INFO: 2016/02/29 17:12:51.704279 docker container: collecting stats for bc59a32f6399f82d90b3db576dab90680b8783bfd38c982d5fd1828141177411
<probe> INFO: 2016/02/29 17:12:51.728818 docker container: collecting stats for 884463e0a22fe0248ab06a7de7f32026a76380145b365e6330dafd0e715ed602
<probe> INFO: 2016/02/29 17:12:51.743856 docker container: collecting stats for 14c1b1338f12a61c8dfd169b1c8bea739ab82d1870eb1c67c6f8961b2530b62e
<probe> INFO: 2016/02/29 17:12:51.772263 docker container: collecting stats for 6b8f060b8d1ada2fa1be3c9fed6f5258cca55a84e544d31c934feb39f46cf476
<probe> ERRO: 2016/02/29 17:12:52.246464 Error collecting weave info, backing off 10s: Get http://127.0.0.1:6784/report: dial tcp 127.0.0.1:6784: getsockopt: connection refused
<probe> ERRO: 2016/02/29 17:13:02.780250 Error collecting weave info, backing off 20s: Get http://127.0.0.1:6784/report: dial tcp 127.0.0.1:6784: getsockopt: connection refused
<probe> ERRO: 2016/02/29 17:13:23.078826 Error collecting weave info, backing off 40s: Get http://127.0.0.1:6784/report: dial tcp 127.0.0.1:6784: getsockopt: connection refused
<probe> ERRO: 2016/02/29 17:14:03.367141 Error collecting weave info, backing off 1m0s: Get http://127.0.0.1:6784/report: dial tcp 127.0.0.1:6784: getsockopt: connection refused
<probe> WARN: 2016/02/29 17:33:02.472231 Topology tagger took longer than 1s
```

```
$ kubectl --kubeconfig=clusters/demo-kubecon-1/kubeconfig logs weave-scope-probe-skgur
<probe> INFO: 2016/02/29 17:12:51.579046 probe starting, version 3e611f8, ID 148e746e80532012
<probe> INFO: 2016/02/29 17:12:51.579122 publishing to: 10.3.0.131:4040
<probe> INFO: 2016/02/29 17:12:51.722956 docker container: collecting stats for 3fc199c5b99f27314ac0d2a80ec8a0603482ea73841a1044289e394affa59526
<probe> INFO: 2016/02/29 17:12:51.762667 docker container: collecting stats for 2f586bcca8f9843bd81b5938296e46419d773a5726f1e81a5c71329cf288745b
<probe> INFO: 2016/02/29 17:12:51.802402 docker container: collecting stats for a98aace0bcb53152e4a104f0d7608567d35a948c7afe693417a65103d22d5c5d
<probe> INFO: 2016/02/29 17:12:51.858837 docker container: collecting stats for 80f5e6a305d4251cf7beebd59fc8d71c99fc291e83d96e125c303147e7e3612c
<probe> INFO: 2016/02/29 17:12:51.914281 docker container: collecting stats for 56e4106a789feb06a78623006f293d30b59a62b471f9bc320b56f934d5b19016
<probe> INFO: 2016/02/29 17:12:51.953305 docker container: collecting stats for 0ce4a1e48fd68507382c3da2bd09b81761bd0ef5604db37087791c85d7a21f69
<probe> INFO: 2016/02/29 17:12:52.063799 docker container: collecting stats for 1885674e5393122b2c8045e49a896b568162f0d3a9644274690f9ef3b1903056
<probe> INFO: 2016/02/29 17:12:52.175927 Control connection to 10.3.0.131:4040 starting
<probe> INFO: 2016/02/29 17:12:52.216306 docker container: collecting stats for bf4be380b5afed2d23b097c8a2d2e811656b5fe661c09c5651a3b8514f19d77b
<probe> INFO: 2016/02/29 17:12:52.211176 Publish loop for 10.3.0.131:4040 starting
<probe> INFO: 2016/02/29 17:12:52.342913 docker container: collecting stats for 53d800d0fe73f005a6d87296528dfc3247a85ea01f09865f3d33570e91df2d78
<probe> INFO: 2016/02/29 17:12:52.383932 docker container: collecting stats for d76e02313508287e92030f729ccc644dd9782e3b4ad5db455ef3a054541b7535
<probe> INFO: 2016/02/29 17:12:52.396967 docker container: collecting stats for 5ad0bbc90cec560d1b1a0a851fb04b7ff12ebbffd9aa626b42757292ed60d64b
<probe> INFO: 2016/02/29 17:12:52.416500 docker container: collecting stats for f8e40ead15cf7e7e47330ec8ba748a38484355b47f6b6d2e05eb37f3a13a163a
<probe> INFO: 2016/02/29 17:12:52.451503 docker container: collecting stats for c70792b76085702c9b90ab532fc8f828b15ac1d639201b543f817b5c3df0c81b
<probe> ERRO: 2016/02/29 17:12:52.756453 Error collecting weave info, backing off 10s: Get http://127.0.0.1:6784/report: dial tcp 127.0.0.1:6784: getsockopt: connection refused
<probe> ERRO: 2016/02/29 17:13:03.134887 Error collecting weave info, backing off 20s: Get http://127.0.0.1:6784/report: dial tcp 127.0.0.1:6784: getsockopt: connection refused
<probe> ERRO: 2016/02/29 17:13:23.431598 Error collecting weave info, backing off 40s: Get http://127.0.0.1:6784/report: dial tcp 127.0.0.1:6784: getsockopt: connection refused
<probe> ERRO: 2016/02/29 17:14:03.814557 Error collecting weave info, backing off 1m0s: Get http://127.0.0.1:6784/report: dial tcp 127.0.0.1:6784: getsockopt: connection refused
```

```
$ kubectl --kubeconfig=clusters/demo-kubecon-1/kubeconfig logs weave-scope-probe-y7iyh
<probe> INFO: 2016/02/29 17:12:42.277550 probe starting, version 3e611f8, ID 3dfa72f7be3cc505
<probe> INFO: 2016/02/29 17:12:42.277664 publishing to: 10.3.0.131:4040
<probe> INFO: 2016/02/29 17:12:42.429918 Control connection to 10.3.0.131:4040 starting
<probe> INFO: 2016/02/29 17:12:42.463385 Publish loop for 10.3.0.131:4040 starting
<probe> INFO: 2016/02/29 17:12:42.464574 docker container: collecting stats for 367baa3d6dcf3228f241e442b4d100498e92a1dd055c0573ef86e6e394e2017e
<probe> INFO: 2016/02/29 17:12:42.528106 docker container: collecting stats for 2b57ff36fdaaff2b2498b76b74ae1050c415381dbb69a747b867249187b3951c
<probe> INFO: 2016/02/29 17:12:42.589344 docker container: collecting stats for 38df1062d9540118fdc0c1e5c4f6b61387a5376d179d9e33209f8a98eed04524
<probe> INFO: 2016/02/29 17:12:42.589518 docker container: collecting stats for 5e683f000a48ae2879c655ae3ed37065cc910395d778ba4ac2fb09157c1b4252
<probe> INFO: 2016/02/29 17:12:42.669804 docker container: collecting stats for 1fda11124b507b5d25b5aec16a003e068ac3df489fdc7560ebbfc04bc2d43378
<probe> INFO: 2016/02/29 17:12:42.709880 docker container: collecting stats for 11497892173fec39355c4c13bbc8e37db4203fae4fe0452a7ecff40baa24363b
<probe> INFO: 2016/02/29 17:12:42.746577 docker container: collecting stats for c7f66b52b817e01fdd3973ac9c77bb4319611bdc9564a0467a7ed800b7dbbd5d
<probe> INFO: 2016/02/29 17:12:42.793396 docker container: collecting stats for dfd970c9024fc35a5d695a9583850a49b1cf7b90b86ccf10cd6b9d829309d928
<probe> INFO: 2016/02/29 17:12:42.807834 docker container: collecting stats for 4f2ed99ec182196c4b58b8316b181815c77ba4f47630f24acbdff1a12d14c02c
<probe> INFO: 2016/02/29 17:12:42.821778 docker container: collecting stats for f352831ee4f868140c9ed8bbf1476c2c813053b24aa30c1c287a60e3cb2b0142
<probe> INFO: 2016/02/29 17:12:42.835938 docker container: collecting stats for 55cab35c1ce5544af5b49053e921e8544315e02c9522dd57a6d01201b49ff862
<probe> INFO: 2016/02/29 17:12:42.855060 docker container: collecting stats for 8f101513fc464716a80dbb6700f2f3167ef04d1e3a1244714b4881f082f24f41
<probe> ERRO: 2016/02/29 17:12:43.067938 Error collecting weave info, backing off 10s: Get http://127.0.0.1:6784/report: dial tcp 127.0.0.1:6784: getsockopt: connection refused
<probe> ERRO: 2016/02/29 17:12:53.411898 Error collecting weave info, backing off 20s: Get http://127.0.0.1:6784/report: dial tcp 127.0.0.1:6784: getsockopt: connection refused
<probe> ERRO: 2016/02/29 17:13:13.847941 Error collecting weave info, backing off 40s: Get http://127.0.0.1:6784/report: dial tcp 127.0.0.1:6784: getsockopt: connection refused
<probe> ERRO: 2016/02/29 17:13:54.121872 Error collecting weave info, backing off 1m0s: Get http://127.0.0.1:6784/report: dial tcp 127.0.0.1:6784: getsockopt: connection refused
```

**Comment by tomwilkie** [comment]
10.3.0.131 is the kubeproxy floating IP in this case.

**Comment by alban** [comment]
It works better when I use AWS EC2 m3.medium instead of t2.micro.

I noticed errors about websocket in CircleCI tests as well:
https://circleci.com/gh/alban/scope/23
```
time="2016-03-01T17:49:03Z" level=info msg="err: websocket: close 1006 (abnormal closure): unexpected EOF" 

--- FAIL: TestCollector (0.01s)
```

**Comment by alban** [comment]
I let Scope running during the night but I suspended my laptop so there should be no new connections. However, the logs are full of errors:
```
$ kubectl --kubeconfig=clusters/demo-kubecon-1/kubeconfig logs weave-scope-app-26t3d |wc -l 
416589
```

```
<app> INFO: 2016/03/01 22:49:47.921254 Error copying to pipe pipe-3726754479046235353 (1) websocket: websocket: close 1006 (abnormal closure): unexpected EOF
<app> INFO: 2016/03/01 22:49:47.923754 Error copying to pipe pipe-3726754479046235353 (1) websocket: websocket: close 1006 (abnormal closure): unexpected EOF
<app> INFO: 2016/03/01 22:49:47.924923 Pipe success pipe-3726754479046235353 (1)
<app> INFO: 2016/03/01 22:49:47.926185 Error copying to pipe pipe-3726754479046235353 (1) websocket: websocket: close 1006 (abnormal closure): unexpected EOF
<app> INFO: 2016/03/01 22:49:47.926501 Pipe success pipe-3726754479046235353 (1)
<app> INFO: 2016/03/01 22:49:47.928285 Error copying to pipe pipe-3726754479046235353 (1) websocket: websocket: close 1006 (abnormal closure): unexpected EOF
<app> INFO: 2016/03/01 22:49:47.937408 Pipe success pipe-3726754479046235353 (1)
<app> INFO: 2016/03/01 22:49:47.938742 Error copying to pipe pipe-3726754479046235353 (1) websocket: websocket: close 1006 (abnormal closure): unexpected EOF
<app> INFO: 2016/03/01 22:49:47.942592 Pipe success pipe-3726754479046235353 (1)
<app> INFO: 2016/03/01 22:49:47.948710 Pipe success pipe-3726754479046235353 (1)
<app> INFO: 2016/03/01 22:49:47.949986 Error copying to pipe pipe-3726754479046235353 (1) websocket: websocket: close 1006 (abnormal closure): unexpected EOF
<app> INFO: 2016/03/01 22:49:47.950116 Error copying to pipe pipe-3726754479046235353 (1) websocket: websocket: close 1006 (abnormal closure): unexpected EOF
<app> INFO: 2016/03/01 22:49:47.951387 Pipe success pipe-3726754479046235353 (1)
<app> INFO: 2016/03/01 22:49:47.952895 Pipe success pipe-3726754479046235353 (1)
<app> INFO: 2016/03/01 22:49:47.953353 Error copying to pipe pipe-3726754479046235353 (1) websocket: websocket: close 1006 (abnormal closure): unexpected EOF
<app> INFO: 2016/03/01 22:49:47.965745 Error copying to pipe pipe-3726754479046235353 (1) websocket: websocket: close 1006 (abnormal closure): unexpected EOF
<app> INFO: 2016/03/01 22:49:47.965916 Timing out pipe pipe-4887636051449339536
<app> INFO: 2016/03/01 22:49:47.966071 Timing out pipe pipe-3726754479046235353
<app> ERRO: 2016/03/01 22:54:56.087890 Error on websocket: websocket: close 1006 (abnormal closure): unexpected EOF
<app> ERRO: 2016/03/01 22:57:15.021503 Error on websocket: websocket: close 1006 (abnormal closure): unexpected EOF
<app> ERRO: 2016/03/01 22:57:15.079442 Error on websocket: websocket: close 1006 (abnormal closure): unexpected EOF
```

Is it possible that Scope does not clean old connections correctly?

**Comment by davkal** [comment]
I tried to focus on the reconnect issue with current `master`. Reconnect for canvas, details panel and terminal worked while using an SSH tunnel to connect to Scope even under following conditions:

* pause + unpause `weavescope` container
* sleep/wakeup laptop

When I restart the `weavescope` container, canvas and details panel reconnects work, but the terminal doesnt (I'm guessing the pipe was kept in memory), and it takes a bit for the UI to realize that the pipe is gone.

The websocket errors in the backend I get when I exit a terminal via `^D`.

**Comment by tomwilkie** [comment]
I tried to reproduce:

```
# kubectl get services
NAME              CLUSTER_IP     EXTERNAL_IP   PORT(S)                      SELECTOR                   AGE
...
weave-scope-app   10.0.248.147   <none>        4040/TCP                     provider=weave-scope-app   44d
#  ssh -L4042:10.0.248.147:4040 -i XXX ubuntu@Y.Y.Y.Y
...
```

And the UI worked fine for me.  This was with 0.13.1

I think the websocket errors are unrelated, and fixed by #1172.
